File: packs/dsa-trees-py/challenges/01-same-tree.test.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    """Build binary tree from level-order list. None means no node."""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root):
    """Convert tree to level-order list (with None for missing nodes)."""
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


from solution import is_same_tree


def test_same_trees():
    p = build_tree([1, 2, 3])
    q = build_tree([1, 2, 3])
    assert is_same_tree(p, q) is True


def test_different_structure():
    p = build_tree([1, 2])
    q = build_tree([1, None, 2])
    assert is_same_tree(p, q) is False


def test_both_empty():
    assert is_same_tree(None, None) is True


def test_one_empty():
    p = build_tree([1])
    assert is_same_tree(p, None) is False


def test_different_values():
    p = build_tree([1, 2, 3])
    q = build_tree([1, 2, 4])
    assert is_same_tree(p, q) is False


def test_single_node_same():
    p = build_tree([1])
    q = build_tree([1])
    assert is_same_tree(p, q) is True

File: packs/dsa-trees-ts/challenges/01-same-tree.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function buildTree(values: (number | null)[]): TreeNode | null {
  if (!values.length || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

function treeToList(root: TreeNode | null): (number | null)[] {
  if (!root) return [];
  const result: (number | null)[] = [];
  const queue: (TreeNode | null)[] = [root];
  while (queue.length) {
    const node = queue.shift()!;
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }
  while (result.length && result[result.length - 1] === null) result.pop();
  return result;
}

const CHALLENGE = join(import.meta.dirname, '01-same-tree.json');

describe('Same Tree', () => {
  let isSameTree: (p: TreeNode | null, q: TreeNode | null) => boolean;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ isSameTree: typeof isSameTree }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    isSameTree = mod.isSameTree;
  });

  afterAll(() => cleanup());

  it('returns true for identical trees', () => {
    expect(isSameTree(buildTree([1, 2, 3]), buildTree([1, 2, 3]))).toBe(true);
  });

  it('returns false for different structure', () => {
    expect(isSameTree(buildTree([1, 2]), buildTree([1, null, 2]))).toBe(false);
  });

  it('returns true for both empty', () => {
    expect(isSameTree(null, null)).toBe(true);
  });

  it('returns false when one is empty', () => {
    expect(isSameTree(buildTree([1]), null)).toBe(false);
  });

  it('returns false for different values', () => {
    expect(isSameTree(buildTree([1, 2, 3]), buildTree([1, 2, 4]))).toBe(false);
  });

  it('returns true for single node same', () => {
    expect(isSameTree(buildTree([1]), buildTree([1]))).toBe(true);
  });
});
